### ISSUE TITLE
Improve `normalizePath` performance

### ISSUE THREAD

**Issue by scolladon** [isomorphic-git/isomorphic-git#1852]
From my benchmark I see the code is spending most of its time in the [`normalizePath`](https://github.com/isomorphic-git/isomorphic-git/blob/ad9bc11e2df311da540330d2243cb3ba9dd5d47a/src/utils/normalizePath.js#L1) function.
When I looked at it I found it is doing multiple Regex in sequence.
This can be optimized with multiple ways

I had a ~800% performance improvement on very large repository using this `normalizePath` implementation:
```javascript
function normalizePath(path) {
  path = path.split('/./').join('/').replace(/\/{2,}/g, '/');
  
  if (path === '/.') return '/';
  if (path === './') return '.';
  if (path.startsWith('./')) path = path.slice(2);
  if (path.endsWith('/.')) path = path.slice(0, -2);
  if (path.length > 1 && path.endsWith('/')) path = path.slice(0, -1)

  return path || '.';
}
```

We could also compare this implementation with a compiled Regex implementation and matching

I think it could be improved further with a very good knowledge of the library.
This is why I do not propose a Pull Request, I think it should be discussed before decided.
What do you think guys ?

**Comment by jcubic** [comment]
It's a good idea, but note that if you do not create a PR probably no one will.
There are also a lot of unit tests so if they pass probably the function is ok.
Sorry, I'm not an expert to see if the function is current or not by looking at the code. I'm also don't know what the old code was doing.

**Comment by scolladon** [comment]
Thanks for your advice @jcubic 
I just created a [PR](https://github.com/isomorphic-git/isomorphic-git/pull/1853) with the performance improvement.
I hope it could be integrated and release soon

**Comment by mojavelinux** [comment]
I would much prefer if we used [path.posix.normalize](https://nodejs.org/api/path.html#pathnormalizepath) when running in Node.js rather than any sort of custom implementation. I don't like the use of a global memorization map. That's a poor practice in Node.js. And this function is already accelerated in Node.js. In the browser, it's not as much of an issue since the script is running per page load (thus naturally scoped).

There are two slight differences when using path.posix.normalize, though I think those could be added as a check in a wrapper. I'd be fine with that.

If we want to proceed, I'd be willing to figure what needs to be done to make use of the function from Node.js when running in that environment.

**Comment by scolladon** [comment]
Hi @mojavelinux, I see your point, I was able to make the test `test-utils-join.js` path with this implementation:
```javascript
const path = require('path').posix || require('path')

export function normalizePath(str) {
  let n = path.normalize(str)

  if (n === './') {
    n = '.'
  }

  return n
}
```

Performance wise it was roughly the same as the implementation proposed (the actual one merged recently, not the previous one)

I looked a bit more deeply how this method is used and why it exist and it seems it is only used by the `join.js` file.
And when looking at the `join.js` file it seems it has been created because `path.posix.join` [seemed to be undefined in webpack at that time](https://github.com/isomorphic-git/isomorphic-git/blob/53310498e9dd84a213197de2d69047ce733580e6/src/utils/join.js#L1).

Maybe we could get rid of both `join.js` and `normalizePath.js` if the `path.posix.join` method with current webpack.
I have not tested that approach yet but I think it will go the way you are suggesting @mojavelinux.
